server/src/main/java/org/elasticsearch/action/StepListener.java

```diff
@@ -92,20 +89,17 @@ public <OtherResponse, OuterResponse> StepListener<OuterResponse> thenCombine(
     }
 
     /**
-     * Returns the future associated with the given step listener
+     * Gets the result of this step. This method will throw {@link IllegalStateException} if this step is not completed yet.
      */
-    public Future<Response> asFuture() {
-        return delegate;
+    public Response result() {
+        return delegate.result();
     }
 
     /**
-     * Gets the result of this step. This method will throw {@link IllegalStateException} if this step is not completed yet.
+     * @return whether this step is complete yet.
      */
-    public Response result() {
-        if (delegate.isDone() == false) {
-            throw new IllegalStateException("step is not completed yet");
-        }
```

Comment:
I think we still need these lines to ensure that the `IllegalStateException` is thrown as specified in the javadocs, otherwise we will hit the assertion in `PlainActionFuture.result()`. Or we should change the javadoc to make it clear that it is illegal (and check usages I suppose).

Reply:
Ah good point, I'll fix the docs here. All the (production) usages look to make sure that the step is done first indeed.